src/GafferScene/MotionPath.cpp

```diff
@@ -0,0 +1,380 @@
+	}
+}
+
+void MotionPath::hashBound( const ScenePath &path, const Context *context, const ScenePlug *parent, MurmurHash &h ) const
+{
+	if( !adjustBoundsPlug()->getValue() )
+	{
+		h = inPlug()->boundPlug()->hash();
+		return;
+	}
+
+	FilteredSceneProcessor::hashBound( path, context, parent, h );
+
+	filterHash( context, h );
+
+	outPlug()->objectPlug()->hash( h );
+	outPlug()->childBoundsPlug()->hash( h );
+}
+
+Imath::Box3f MotionPath::computeBound( const ScenePath &path, const Context *context, const ScenePlug *parent ) const
+{
+	if( !adjustBoundsPlug()->getValue() )
+	{
+		return inPlug()->boundPlug()->getValue();
+	}
+
+	const PathMatcher::Result m = filterValue( context );
+	if( m & ( PathMatcher::ExactMatch | PathMatcher::DescendantMatch ) )
+	{
+		Imath::Box3f result = outPlug()->childBoundsPlug()->getValue();
+		if( m & PathMatcher::ExactMatch )
+		{
+			if( ConstCurvesPrimitivePtr motionPath = runTimeCast<const CurvesPrimitive>( outPlug()->objectPlug()->getValue() ) )
+			{
+				result.extendBy( SceneAlgo::bound( motionPath.get() ) );
+			}
+		}
+		return result;
+	}
+
+	return outPlug()->childBoundsPlug()->getValue();
+}
+
+void MotionPath::hashTransform( const ScenePath &path, const Context *context, const ScenePlug *parent, MurmurHash &h ) const
+{
+	h = MurmurHash();
+	h.append( inPlug()->transformPlug()->defaultValue() );
+}
+
+Imath::M44f MotionPath::computeTransform( const ScenePath &path, const Context *context, const ScenePlug *parent ) const
+{
+	return inPlug()->transformPlug()->defaultValue();
+}
+
+void MotionPath::hashAttributes( const ScenePath &path, const Context *context, const ScenePlug *parent, MurmurHash &h ) const
+{
+	h = inPlug()->attributesPlug()->defaultValue()->Object::hash();
+}
+
+ConstCompoundObjectPtr MotionPath::computeAttributes( const ScenePath &path, const Context *context, const ScenePlug *parent ) const
+{
+	return inPlug()->attributesPlug()->defaultValue();
+}
+
+void MotionPath::hashObject( const ScenePath &path, const Context *context, const ScenePlug *parent, MurmurHash &h ) const
+{
+	FilteredSceneProcessor::hashObject( path, context, parent, h );
+
+	filterHash( context, h );
+
+	h.append( inPlug()->fullTransformHash( path ) );
+
+	h.append( context->getFrame() );
+
+	startModePlug()->hash( h );
+	startFramePlug()->hash( h );
+	endModePlug()->hash( h );
+	endFramePlug()->hash( h );
+	incrementPlug()->hash( h );
+}
+
+ConstObjectPtr MotionPath::computeObject( const ScenePath &path, const Context *context, const ScenePlug *parent ) const
+{
+	if( path.empty() || !(filterValue( context ) & PathMatcher::ExactMatch) )
+	{
+		return inPlug()->objectPlug()->defaultValue();
+	}
+
+	float start = ( (FrameMode)startModePlug()->getValue() == FrameMode::Absolute ) ? startFramePlug()->getValue() : context->getFrame() + startFramePlug()->getValue();
+	float end = ( (FrameMode)endModePlug()->getValue() == FrameMode::Absolute ) ? endFramePlug()->getValue() : context->getFrame() + endFramePlug()->getValue();
```

Comment:
For the first bullet, your concern is floating point error, because we're doing a floating point loop that could be right on the edge of how many iterations it fits in?  I guess this could also result in `steps` not matching the number of loop iterations, based on floating point error?  It seems like the solution would be to compute `steps` based on the raw start/end before potentially adding the frame number, and then have the loop index just be integer 0 ... steps - 1, with an appropriate conversion from index to time.

It might be good to use an epsilon to avoid doubling up the end vertex?  Currently, it's exactly down to floating point error whether the end vertex gets duped?

For the second bullet, motion blur on the trail would sometimes be desirable.  If you don't want it, maybe it's OK to use a separate TimeWarp, though we should probably have an easy checkbox for doing "warp to nearest frame" on TimeWarp?

I also wonder if we should make any effort to sample on frame.  If someone wants to create a bidirectional trail that grows in length, so it starts out being -2 .. 2, and grows to -4 .. 4, with an increment of 1, then when it is from -2.5 .. 2.5, is it weird that it takes all it's samples in between frames?

Reply:
> For the first bullet, your concern is floating point error, because we're doing a floating point loop that could be right on the edge of how many iterations it fits in?

No, I was _separately_ worrying about floating point error, and especially how we might do more/less iterations than we have in `steps`. But since we `reserve()` the buffer and `emplace_back()` rather than `resize()` I decided not to worry too much.

The mismatched samples case is much simpler - a trail with start of absolute 0 and end of relative 0 will vary in length over time. Depending on your increment, that can give mismatched samples within a frame.

Reply:
> For the second bullet, motion blur on the trail would sometimes be desirable. If you don't want it, maybe it's OK to use a separate TimeWarp, though we should probably have an easy checkbox for doing "warp to nearest frame" on TimeWarp?

If you don't want motion blur, it's easy to just turn it off using StandardAttributes, no need for a TimeWarp. But it will be non-obvious to people that that's the solution to the weird errors about mismatched sample lengths they might be getting. What are the use cases you're imagining where you do want motion blur?

Reply:
I guess the motion graphics cases where you do want want motion blur aren't terribly useful in Gaffer yet ( and maybe they would apply more to a vertex trail node than a location trail node ).  But I don't feel like we should completely rule out the possibility of wanting a magic swoosh that follows a location around and actually gets rendered as a visual effect, in which case you would want it to motion blur properly.

Reply:
I added a test to mimic motion blur sampling in a5ee7c1128eb4991bf1ac2ae06b47b92622d1e89. Take a close look and see if it makes sense... That test was definitely failing prior to this commit, but I'm not 100% confident in the new logic. Dan suggests there is still a bug that none of my tests are catching and we should need to compute `steps` prior to adding current frame to `start`/`end`. I'll try to repro that issue first before making the code more complex though.

Reply:
To clarify, the commit aims to address the concerns about duplicating the end point due to floating point error and ensuring the vert count stays constant across a shutter in the “normal” case where a user wanted a consistently sampled curve.

My opinions on the other points are:

- As Dan says, rendering the curve as a visual effect can be desirable, so we shouldn’t force-disable motion blur, we should leave it up to the user via StandardAttributes.
- Explicit sampling on integer frames is unnecessary, we sample geometry off frame or misaligned with an exact keyframe all the time, and it’s almost never an issue in practice.
- When the user intentionally chooses a changing-number-of-samples curve (via mixing absolute and relative modes or animating the start/end frames), it’s on them to debug messages about changing vert counts at rendertime.

Reply:
> But I don't feel like we should completely rule out the possibility of wanting a magic swoosh that follows a location around and actually gets rendered as a visual effect, in which case you would want it to motion blur properly.

I don't understand what you are wanting from motion blur in this case. The complete trail for an entire shot length is fixed in space, and all the `start/end/increment` parameters do is let you choose a span within it. Any frame-to-frame movement comes from changes to those parameters or the current frame, and can only add/remove bits from the end of the trail, or change the sampling within it. Either that breaks motion blur by changing the number of vertices, or gives a fuzzy line as the sample points jitter in terms of the exact time they sample the curve. Neither of those seem particularly desirable to me. Perhaps if we added an option to use a fixed number of vertices rather than an increment then motion blur might be more useful?

Here's an example of the jitter I'm talking about - a sine wave where vertices at different time samples choose different parts of the curve :

![image](https://user-images.githubusercontent.com/1133871/106874857-f59d7d00-66cd-11eb-9139-fbf86fc42a35.png)

> Explicit sampling on integer frames is unnecessary, we sample geometry off frame or misaligned with an exact keyframe all the time, and it’s almost never an issue in practice.

Integer frames aren't a concern to me, but I think there might be some merit in fixing the samples at exact multiples of the increment. That could solve the jitter problem above, and I think could also be a step towards solving the mismatched vertices issue.

> When the user intentionally chooses a changing-number-of-samples curve (via mixing absolute and relative modes or animating the start/end frames), it’s on them to debug messages about changing vert counts at rendertime.

I could accept this in a first version, but I think it'll cause confusion. Absolute start 0 and relative end 0 are a pretty likely choice of settings that lead directly into this trap.

Reply:
> Here's an example of the jitter I'm talking about - a sine wave where vertices at different time samples choose different parts of the curve

Sure, that sort of jittery motion graphics effect might be desirable for _something_, but it would be easier to control it with a noise deformer that gives you precise control over modifications to a stable underlying curve.

Reply:
Summary of our discussion:

- The current incremental/step sampling is good for previewing a non-blurable motion path at regular intervals
- The motion graphics case merits a different sampling mode to ensure the vert count is constant during a render
- We'll add `samplingMode` with `Fixed` and `Variable` options. `Variable` mode will use the float `increment` plug (renamed to `step` for user-friendliness), while `Fixed` mode will use a new int `samples` plug (min value 2).
- We won't adjust attributes in either mode, so its on the user to disable deformation blur when they need to. We'll add a tooltip to clarify that point.

Reply:
Addressed in e097c4d